Refactor mm processors and Enable mixed modality processing

JustinTong0323 opened the pull request:
<!-- Thank you for your contribution! We appreciate it. The following guidelines will help improve your pull request and facilitate feedback. If anything is unclear, don't hesitate to submit your pull request and ask the maintainers for assistance. -->

## Motivation

Refactors the multimodal processor to support mixed modalities (images and audio) in the same request.

It introduces a unified processing flow that handles different input types, reducing code duplication and improving maintainability. The new design allows creating multimodal data items directly from processor outputs, streamlining the data handling process.

<!-- Explain the purpose of this PR and the goals it aims to achieve. -->

## Modifications



<!-- Describe the changes made in this PR. -->

## Checklist

- [x] Format your code according to the [Code Formatting with Pre-Commit](https://docs.sglang.ai/references/contribution_guide.html#code-formatting-with-pre-commit).
- [ ] Add unit tests as outlined in the [Running Unit Tests](https://docs.sglang.ai/references/contribution_guide.html#running-unit-tests-adding-to-ci).
- [ ] Update documentation / docstrings / example tutorials as needed, according to [Writing Documentation](https://docs.sglang.ai/references/contribution_guide.html#writing-documentation-running-docs-ci).
- [ ] Provide throughput / latency benchmark results and accuracy evaluation results as needed, according to [Benchmark and Profiling](https://docs.sglang.ai/references/benchmark_and_profiling.html) and [Accuracy Results](https://docs.sglang.ai/references/accuracy_evaluation.html).
- [ ] For reviewers: If you haven't made any contributions to this PR and are only assisting with merging the main branch, please remove yourself as a co-author when merging the PR.
- [ ] Please feel free to join our Slack channel at https://slack.sglang.ai to discuss your PR.

JustinTong0323 committed:
Enables mixed modality processing

Refactors the multimodal processor to support mixed modalities (images and audio) in the same request.

It introduces a unified processing flow that handles different input types, reducing code duplication and improving maintainability.
The new design allows creating multimodal data items directly from processor outputs, streamlining the data handling process.

Signed-off-by: Xinyuan Tong <justinning0323@outlook.com>

mickqian commented on python/sglang/srt/models/gemma3n_mm.py:
```diff
@@ -252,11 +252,11 @@ def pad_input_ids(
+            media_token_pairs.append((mm_inputs.im_start_id, mm_inputs.im_end_id))
+        if mm_inputs.audio_start_id is not None:
+            media_token_pairs.append((mm_inputs.audio_start_id, mm_inputs.audio_end_id))
+        print(f"DEBUG: gemma3n_mm.pad_input_ids: {media_token_pairs=}")
```
We should consider manage all these special tokens into one `MultimodalSpecialTokens`. When building it, `chat template` of this model can be considered as a reference, as it already contains some special tokens.

mickqian commented on python/sglang/srt/multimodal/processors/base_processor.py:
```diff
@@ -109,6 +100,30 @@ def __init__(self, hf_config, server_args, _processor):
+            "input_features": Modality.AUDIO,
+            "input_features_mask": Modality.AUDIO,
+            # Video-related attributes
+            "video_grid_thws": Modality.VIDEO,
```
Maybe it's more appropriate if we pass this dict model-specifically?

mickqian replied:
we should simplify this, only real data attributes are needed

mickqian replied:
We can keep the logic in parent class, while letting subclass decides the actual modality by providing a mapping. This big dict works well in most scenarios, but still it's a bit hack.

mickqian commented on python/sglang/srt/multimodal/processors/base_processor.py:
```diff
@@ -417,175 +435,133 @@ def _extract_processor_features_from_all_attributes(
+                    items[modality] = MultimodalDataItem(modality=modality)
+
+                # Set attribute
+                if hasattr(items[modality], attr_name):
```
we can skip this judgement?

mickqian commented on python/sglang/srt/multimodal/processors/base_processor.py:
```diff
@@ -417,175 +435,133 @@ def _extract_processor_features_from_all_attributes(
                 values[k] = v
         return values
 
+    def _create_mm_items_from_dict(self, data_dict: dict) -> List[MultimodalDataItem]:
```
`collect_mm_items_from_processor_output`

mickqian commented on python/sglang/srt/multimodal/processors/base_processor.py:
```diff
@@ -417,175 +435,133 @@ def _extract_processor_features_from_all_attributes(
+
+        return list(items.values())
+
+    def _process_and_create_mm_items(
```
`_process_and_collect_mm_items`

mickqian commented on python/sglang/srt/multimodal/processors/base_processor.py:
```diff
@@ -417,175 +435,133 @@ def _extract_processor_features_from_all_attributes(
+        """Create mm_items directly from processor output."""
+        items = {}  # modality -> MultimodalDataItem
+
+        for attr_name, value in data_dict.items():
```
As above, decide the modality from whether the real data attribute is presented in this dict

mickqian commented on python/sglang/srt/multimodal/processors/base_processor.py:
```diff
@@ -417,175 +435,133 @@ def _extract_processor_features_from_all_attributes(
+            return [], input_ids
+
+        dict_items, raw_images, raw_audios = [], [], []
+        for item in all_items:
```
TODO: the modality of the precomputed features will have to be passed along with data in the future

JustinTong0323 committed:
Simplifies multimodal token padding

Refactors the multimodal token padding logic to directly replace specific multimodal tokens (image, audio, video) with their corresponding padding values.

This change replaces the previous approach of identifying contiguous regions based on start and end tokens with a more straightforward method of replacing individual tokens. It streamlines the padding process and avoids potential issues with mismatched region counts.

Signed-off-by: Xinyuan Tong <justinning0323@outlook.com>

JustinTong0323 committed:
Refactors multimodal processor input handling

Simplifies multimodal processor logic by ensuring image and audio data are consistently handled as lists.

Removes redundant checks for empty or single-element inputs within individual processor implementations, promoting code reuse and maintainability.

Signed-off-by: Xinyuan Tong <justinning0323@outlook.com>

JustinTong0323 committed:
Refactors VILA multimodal processing

Refactors the VILA multimodal processing logic for better organization
and efficiency. It streamlines the data flow and adds explicit handling
of image and video token IDs for improved clarity and future
extensibility.

Signed-off-by: Xinyuan Tong <justinning0323@outlook.com>

JustinTong0323 committed:
resolve comments: Renames methods for clarity

Renames `_create_mm_items_from_dict` to
`collect_mm_items_from_processor_output` and
`_process_and_create_mm_items` to
`_process_and_collect_mm_items` to improve code readability
and reflect their function more accurately.

Signed-off-by: Xinyuan Tong <justinning0323@outlook.com>

JustinTong0323 committed:
fix lint

Signed-off-by: Xinyuan Tong <justinning0323@outlook.com>

JustinTong0323 committed:
Merge branch 'main' into refactor-base-processor

JustinTong0323 committed:
Fixes audio processing for Gemma3n

Ensures 'audio' is only added to kwargs for Gemma3n processors.

This resolves an issue where the 'audio' keyword argument was being incorrectly added to the keyword arguments for other multimodal processors, causing compatibility issues.

Signed-off-by: Xinyuan Tong <justinning0323@outlook.com>

JustinTong0323 committed:
fix: import Gemma3nSGLangProcessor

Signed-off-by: Xinyuan Tong <justinning0323@outlook.com>

JustinTong0323 committed:
fix last commit

Signed-off-by: Xinyuan Tong <justinning0323@outlook.com>

JustinTong0323 commented:
Current mm data flow in processor is shown as:
```mermaid
graph TD
    subgraph "Phase 1: Loading Data"
        direction LR
        Start((Start)) --> UserInput["User Input <br> (prompt, image_data, audio_data)"];
        UserInput --> LoadMMData["load_mm_data"];
        LoadMMData --> SplitPrompt["Split prompt by special tokens <br> e.g., <image>"];
        SplitPrompt --> SubmitTasks["submit_data_loading_tasks"];
        SubmitTasks --> ParallelLoad["Parallel Loading (IO Thread Pool) <br> _load_single_item loads each file"];
        ParallelLoad --> ReconstructText["Reconstruct text prompt & <br> Collect loaded raw data (PIL Images, etc.)"];
        ReconstructText --> BaseOutput["BaseMultiModalProcessorOutput <br> (text with special tokens, raw images, raw audios)"];
    end

    subgraph "Phase 2: Processing and Combining"
        direction LR
        BaseOutput --> ProcessAndCombine["process_and_combine_mm_data"];
        ProcessAndCombine --> CategorizeData{"Categorize data <br> pre-processed vs raw"};
        CategorizeData -- "Raw data" --> HFProcessor["process_mm_data <br> (Calls HuggingFace processor)"];
        HFProcessor --> ProcessorOutput["Processor Output <br> (input_ids, pixel_values, etc.)"];
        CategorizeData -- "Pre-processed data" --> CollectItems["collect_mm_items_from_processor_output"];
        ProcessorOutput --> CollectItems;
        CollectItems --> CreateMMItems["Create MultimodalDataItem objects <br> with processed features"];
        CreateMMItems --> AddOffsets["Find special token positions in input_ids <br> and add offsets to MultimodalDataItems"];
        AddOffsets --> FinalOutput([Final Output <br> List of MultimodalDataItem, <br> input_ids tensor]);
        FinalOutput --> End((End));
    end
```

JustinTong0323 committed:
Improves Modality.from_str robustness

Ensures `Modality.from_str` is case-insensitive and raises a ValueError with a helpful message when an invalid modality string is provided. This improves the user experience by providing more informative error messages.

Signed-off-by: Xinyuan Tong <justinning0323@outlook.com>

JustinTong0323 commented:
/gemini review

JustinTong0323 commented on python/sglang/srt/managers/mm_utils.py:
```diff
@@ -129,67 +129,34 @@ def pad_input_tokens(
-                output_ids_tensor[start_idx:end_idx] = pad_value
+        for item in mm_inputs.mm_items:
+            if item.is_image() and mm_inputs.im_token_id is not None:
+                token_to_pad_mapping[mm_inputs.im_token_id] = item.pad_value
+            elif item.is_audio() and mm_inputs.audio_token_id is not None:
+                token_to_pad_mapping[mm_inputs.audio_token_id] = item.pad_value
+            elif item.is_video() and mm_inputs.video_token_id is not None:
+                token_to_pad_mapping[mm_inputs.video_token_id] = item.pad_value
             else:
-                logger.warning(f"Skipping region {i} due to None pad_value.")
-        return output_ids_tensor.tolist()
+                raise ValueError(f"No multimodal token id provided for {item.modality}")
```
We assume only one mm_item for each modality, as we already combined.

JustinTong0323 committed:
Simplifies conditional checks for multimodal data

Replaces checks `if images is not None` with `if images` for conciseness.
This change ensures that multimodal data processing logic correctly handles empty lists/None values for images, videos, and audios.

Signed-off-by: Xinyuan Tong <justinning0323@outlook.com>

JustinTong0323 committed:
Merge branch 'main' into refactor-base-processor

JustinTong0323 committed:
Simplifies multimodal token padding

Refactors the multimodal token padding mechanism by removing the need to explicitly pass token IDs to the `MultiModalityDataPaddingPatternMultimodalTokens` class.

This change streamlines the padding process and makes it more consistent across different models.

Signed-off-by: Xinyuan Tong <justinning0323@outlook.com>
